prune/prune.go

```diff
@@ -16,184 +16,204 @@ package prune
+	New Auto Pruning API Implementation
+	-------------------------------------------
+*/
+
+// Pruner is an object that runs a prune job.
+type Pruner struct {
+	// Registry is the IsPrunableFunc registry
+	Registry *Registry
+
+	// Client is the k8s client that will be used
+	Client client.Client
+
+	// DryRun indicates whether or not we should actually perform pruning or just return the list of pruneable objects
+	// true = Just check, don't prune
+	// false (default) = Prune
+	DryRun bool
+
+	// Strategy is the function used to determine a list of resources that are pruneable
+	Strategy StrategyFunc
 
-// StrategyConfig holds settings unique to each pruning mode
-type StrategyConfig struct {
-	Mode            Strategy
-	MaxAgeSetting   string
-	MaxCountSetting int
-	CustomSettings  map[string]interface{}
+	// Labels is a map of the labels to use for label matching when looking for resources
+	Labels map[string]string
+
+	// Namespace is the namespace to use when looking for resources
+	Namespace string
+
+	// Logger is the logger to use when running pruning functionality
+	Logger logr.Logger
 }
 
-// StrategyFunc function allows a means to specify
-// custom prune strategies
-type StrategyFunc func(ctx context.Context, cfg Config, resources []ResourceInfo) ([]ResourceInfo, error)
-
-// PreDelete function is called before a resource is pruned
-type PreDelete func(ctx context.Context, cfg Config, something ResourceInfo) error
-
-// Config defines a pruning configuration and ultimately
-// determines what will get pruned
-type Config struct {
-	Clientset      kubernetes.Interface      // kube client used by pruning
-	LabelSelector  string                    //selector resources to prune
-	DryRun         bool                      //true only performs a check, not removals
-	Resources      []schema.GroupVersionKind //pods, jobs are supported
-	Namespaces     []string                  //empty means all namespaces
-	Strategy       StrategyConfig            //strategy for pruning, either age or max
-	CustomStrategy StrategyFunc              //custom strategy
-	PreDeleteHook  PreDelete                 //called before resource is deleteds
-	Log            logr.Logger               //optional: to overwrite the logger set at context level
+// ErrUnprunable indicates that it is not allowed to prune a specific object.
+type ErrUnprunable struct {
+	Obj    *client.Object
+	Reason string
 }
 
-// Execute causes the pruning work to be executed based on its configuration
-func (config Config) Execute(ctx context.Context) error {
-	log := Logger(ctx, config)
-	log.V(1).Info("Execute Prune")
+// Error returns a string representation of an `ErrUnprunable` error.
+func (e *ErrUnprunable) Error() string {
+	return fmt.Sprintf("%s -- %+v", e.Reason, e.Obj)
+}
+
+// StrategyFunc takes a list of resources and returns the subset to prune.
+type StrategyFunc func(ctx context.Context, objs []client.Object) ([]client.Object, error)
```

Comment:
Can you see if `[]client.ObjectKey` is sufficient for the first return type?

Reply:
It looks like we should be able to use `[]client.ObjectKey` as the return type, but will have to create an object with a namespace, name, and GVK set like so (code taken from https://github.com/kubernetes-sigs/controller-runtime/blob/0a4d154ec52c0359f4b20ec9e2845a2e7af125a0/pkg/client/example_test.go#L221-L230):
```go
	// Using a unstructured object.
	u := &unstructured.Unstructured{}
	u.SetName("name")
	u.SetNamespace("namespace")
	u.SetGroupVersionKind(schema.GroupVersionKind{
		Group:   "apps",
		Kind:    "Deployment",
		Version: "v1",
	})
	_ = c.Delete(context.Background(), u)
```

I am going to take a shot at it and see if it will work how we are expecting.